crates/bevy_render/src/render_resource/buffer_vec.rs

```diff
@@ -20,49 +22,56 @@ impl<T: Pod> Default for BufferVec<T> {
             buffer: None,
             capacity: 0,
             buffer_usage: BufferUsages::all(),
-            item_size: std::mem::size_of::<T>(),
         }
     }
 }
 
 impl<T: Pod> BufferVec<T> {
+    /// Creates a new [`BufferVec`] with the associated [`BufferUsages`].
+    ///
+    /// This does not immediately allocate a buffer.
     pub fn new(buffer_usage: BufferUsages) -> Self {
         Self {
             buffer_usage,
             ..Default::default()
         }
     }
 
+    /// Gets the reference to the underlying buffer, if one has been allocated.
     #[inline]
     pub fn buffer(&self) -> Option<&Buffer> {
         self.buffer.as_ref()
     }
 
+    /// Gets the capacity of the underlying buffer.
+    ///
+    /// Will return 0 if no buffer has been allocated yet.
     #[inline]
-    pub fn capacity(&self) -> usize {
+    pub fn buffer_capacity(&self) -> usize {
         self.capacity
     }
 
-    #[inline]
-    pub fn len(&self) -> usize {
-        self.values.len()
-    }
-
-    #[inline]
-    pub fn is_empty(&self) -> bool {
-        self.values.is_empty()
-    }
-
-    pub fn push(&mut self, value: T) -> usize {
-        let index = self.values.len();
-        self.values.push(value);
-        index
+    /// Queues up a copy of the contents of the [`BufferVec`] into the underlying
+    /// buffer.
+    ///
+    /// If no buffer has been allocated yet or if the current size of the contents
+    /// exceeds the size of the underlying buffer, a new buffer will be allocated.
+    pub fn write_buffer(&mut self, device: &RenderDevice, queue: &RenderQueue) {
```

Comment:
Ah, so the buffer itself will only ever be written to with this method, so allowing free access to the underlying vector is logically-safe?

Reply:
There are the unsafe methods on Vec, which can lead to some invalid state, but IMO it's incumbent on the caller to make that decision to make sure those uses are safe.

Any Vec API that consumes the underlying Vec also cannot be used as it's operating on a reference.

Any other safe API for the Vec, combined with the limitation of only having write_buffer as the sole way to write out to the underlying buffer should make exposing these Vec APIs safe (within reason).